subprojects/core/src/main/java/org/gradle/api/internal/file/collections/ImmutableFileCollection.java

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.file.collections;
+
+import com.google.common.collect.ImmutableSet;
+import org.gradle.api.file.FileCollection;
+import org.gradle.api.internal.file.AbstractFileCollection;
+import org.gradle.api.internal.file.IdentityFileResolver;
+import org.gradle.internal.file.PathToFileResolver;
+
+import java.io.File;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.LinkedHashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Set;
+
+public class ImmutableFileCollection extends AbstractFileCollection {
+    private static final String DEFAULT_DISPLAY_NAME = "immutable file collection";
+    private static final EmptyImmutableFileCollection EMPTY = new EmptyImmutableFileCollection();
+
+    private final Set<Object> files;
+    private final String displayName;
+    private final PathToFileResolver resolver;
+
+    public static ImmutableFileCollection of(Object... files) {
+        return of(Arrays.asList(files));
+    }
+
+    public static ImmutableFileCollection of(Collection<?> files) {
+        return usingResolver(new IdentityFileResolver(), files);
+    }
+
+    public static ImmutableFileCollection usingResolver(PathToFileResolver fileResolver, Object[] files) {
+        return usingResolver(fileResolver, Arrays.asList(files));
+    }
+
+    public static ImmutableFileCollection usingResolver(PathToFileResolver fileResolver, Collection<?> files) {
+        if (files.isEmpty()) {
+            return EMPTY;
+        }
+        return new ImmutableFileCollection(DEFAULT_DISPLAY_NAME, fileResolver, files);
+    }
+
+    private ImmutableFileCollection(String displayName, PathToFileResolver fileResolver, Collection<?> files) {
```

Comment:
We should use `Iterable` instead of `Collection`. This allows passing things like `Iterables.concat()` etc., and we don't actually need a full-blown `Collection` here.